### PR TITLE
Tests | Fix Tests regarding Concurrency both Sync and Async versions.

### PR DESCRIPTION
**Pull request by H-Yeo**
On Bug Ticket #25147, we are seeing intermittent failure on the ConcurrentExecution tests.

After analysis, we have determined that as the tests are running in the pipeline, we are witnessing thread starvation as the tests are running. In addition, there is also an edge case where there an exception before the retry mechanism is encountered due to connectivity issues with the database.

In order to catch all exceptions, some refactoring was performed for thread starvation, async/sync test separation, and counting of retry exception occurrences.

### CODE REVIEW

**Comment by H-Yeo** [comment]
All PR Check's Pass in Draft

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs [hunk 1]
```diff
@@ -524,160 +526,164 @@ public async void RetryExecuteAsyncCancel(string cnnString, SqlRetryLogicBasePro
...
-                        cmd.CommandText = query + " FOR XML AUTO";
-                        Assert.Throws<AggregateException>(() => cmd.ExecuteXmlReader());
+                        cmd.CommandText = query;
+
+                        for (int j = 0; j < delayCount; j++)
+                        {
+                            int longQueryOperation = j * j;
+                        }
+
+                        Assert.Throws<Exception>(() => commandAction(cmd));
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs:609]: `AggregateException` is expected in all the tests. I don't see any reason to use the `Exception` instead of it and adding one level of assertion.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs [hunk 2]
```diff
@@ -524,160 +526,164 @@ public async void RetryExecuteAsyncCancel(string cnnString, SqlRetryLogicBasePro
         [MemberData(nameof(RetryLogicTestHelper.GetConnectionAndRetryStrategyInvalidCommand), parameters: new object[] { 2 }, MemberType = typeof(RetryLogicTestHelper), DisableDiscoveryEnumeration = true)]
         public void ConcurrentExecution(string cnnString, SqlRetryLogicBaseProvider provider)
         {
-            int numberOfTries = provider.RetryLogic.NumberOfTries;
             string query = "SELECT bad command";
-            int retriesCount = 0;
             int concurrentExecution = 3;
-            provider.Retrying += (s, e) => Interlocked.Increment(ref retriesCount);
+            ProcessDataInParallel(cnnString, provider, query, concurrentExecution, cmd => cmd.ExecuteScalar());
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs:531]: More tidier. Nice!

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs [hunk 3]
```diff
@@ -524,160 +526,164 @@ public async void RetryExecuteAsyncCancel(string cnnString, SqlRetryLogicBasePro
...
+                        using (SqlCommand cmd = cnn.CreateCommand())
+                        {
+                            await cnn.OpenAsync();
+                            cmd.RetryLogicProvider = provider;
+                            cmd.CommandText = query;
+
+                            for (int j = 0; j < delayCount; j++)
+                            {
+                                int longQueryOperation = j * j;
+                            }
```
[the remark this answers is not in the document]
Comment (H-Yeo) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs:661]: Please LMK if you want this removed, but this was added to reduce/test whether we were encountering Thread Starvation with multiple simultaneous connections to the database. Adding a long query was a potential to reduce connection errors.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs (hunk 1 above)
Comment (H-Yeo) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs:609]: This was a problem with Thread Starvation. The first exception being thrown is coming out of the connection (System.Exception), which is well before we attempt our actual testing of the Retry, which is the expected Aggregate.Exception.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs (hunk 2 above)
Comment (H-Yeo) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs:531]: Thanks!

**Comment by H-Yeo** [comment]
Hi @David-Engel , Would you have able to perform a PR on this when you have a moment. It was revolving around the Thread Starvation issue around connections. LMK if you have any questions. Thanks!

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs:661]: Making the CPU busy is not a safe way to add delay and you can use `Thread.Sleep` for a more stable delay. I don't suggest adding delay here to make it more possible to be in conflict for the underlaying feature implementation. If the connection status is your concern here you can take it out and pass it after making the connection to the server.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs (hunk 1 above)
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs:609]: Just wondering if you have tried without the for-loop delay in the pipeline. I feel this may cause a high-CPU usage. Have you tried Thread.Sleep()?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (David-Engel) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs:661]: I don't understand this block. It looks like it is just using a thread to perform a bunch of CPU operations here to make the async method execute a bit longer. How is this helping identify thread starvation?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs [hunk 4]
```diff
@@ -524,160 +526,164 @@ public async void RetryExecuteAsyncCancel(string cnnString, SqlRetryLogicBasePro
...
+        private static void ProcessDataInParallel(string cnnString, SqlRetryLogicBaseProvider provider,
+                                                  string query, int concurrentExecution,
+                                                  Action<SqlCommand> commandAction)
+        {
+            int numberOfTries = provider.RetryLogic.NumberOfTries;
+            int delayCount = 10_000;
+            int retriesCount = 0;
+            var exceptions = new ConcurrentQueue<Exception>();
+
+            ThreadPool.SetMinThreads(concurrentExecution * 2, concurrentExecution * 2);
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs:590]: I don't think this is the best place to SetMinThreads. SetMinThreads affects the entire process and we need to return it to its previous value to avoid affecting subsequent tests. I'd put SetMinThreads in a constructor for the test class, make the class IDisposable, and set min threads back in the Dispose() method.

Same for line 641.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (H-Yeo) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs]: These long running query operations will be removed, as you're correct, they don't do anything to address the underlying problem.

[the remark this answers is not in the document]
Comment (H-Yeo) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs]: Adjusting the flow of Exception handling as recommend by Davoud.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs (hunk 1 above)
Comment (H-Yeo) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs]: Adjusting the exception handling as mentioned earlier. Will test shortly via Pipeline.

**Review by David-Engel — APPROVED** [review]

**Review by DavoudEshtehari — APPROVED** [review]
